File: algorithms/hard/n_queens_2.py
class Solution:
    def totalNQueens(self, n: int) -> int:
        def solve(r, cset, dset, adset):
            print(r, cset, dset, adset)
            if r == n:
                self.res += 1
                return
            for c in range(n):
                pos = (r, c)
                if c not in cset and r - c not in dset and r + c not in adset:
                    cset.add(c)
                    dset.add(r - c)
                    adset.add(r + c)
                    res = solve(r + 1, cset, dset, adset)
                    if not res:
                        cset.remove(c)
                        dset.remove(r - c)
                        adset.remove(r + c)
            return False
        self.res = 0
        cset, dset, adset = set(), set(), set()
        solve(0, cset, dset, adset)
        return self.res

# Main section
for n in [
            #1,
            #2,
            #3,
            4,
            #5,
            #6,
            #7,
            #8,
            #9,
         ]:
    print(f'n = {n}')
    sol = Solution()
    r = sol.totalNQueens(n)
    print(f'r = {r}')
    print('=================')

File: algorithms/hard/n_queens_2.go
//
// I am unable to make it work. Not sure why the map does not get cleared!
//
package main

import "fmt"

var res int

func totalNQueens(n int) int {
    res = 0
    r := 0
    cset := make(map[int]bool)
    dset := make(map[int]bool)
    adset := make(map[int]bool)
    backtrack(r, n, cset, dset, adset)
    return res
}

func backtrack(r, n int, cset, dset, adset map[int]bool) bool {
    //fmt.Println(r, n, cset, dset, adset)
    if r == n {
        res++
        return true
    }
    for c := 0; c < n; c++ {
        // pos = (r, c)
        _, ok := cset[c]; if ok {continue}
        _, ok = dset[r - c]; if ok {continue}
        _, ok = adset[r + c]; if ok {continue}
        cset[c] = true
        dset[r - c] = true
        adset[r + c] = true
        res := backtrack(r + 1, n, cset, dset, adset)
        if !res {
            delete(cset, c)
            delete(dset, r - c)
            delete(adset, r + c)
        }
    }
    return false
}

func main() {
    data := []int{1,2,3,4,5,6,7,8,9}
    for _, n := range data {
        fmt.Printf("n = %d\n", n)
        totalNQueens(n)
        fmt.Printf("res = %d\n", res)
        fmt.Println("====================")
    }
}


